libnavigation-core/src/main/java/com/mapbox/navigation/core/reroute/RerouteController.kt

```diff
@@ -0,0 +1,78 @@
+package com.mapbox.navigation.core.reroute
+
+import com.mapbox.api.directions.v5.models.DirectionsRoute
+import com.mapbox.navigation.core.MapboxNavigation
+import com.mapbox.navigation.core.trip.session.OffRouteObserver
+
+/**
+ * Reroute controller allow change reroute logic externally. Use [MapboxNavigation.setRerouteController]
+ * to replace that logic.
+ */
+interface RerouteController {
+
+    /**
+     * State of Reroute process
+     */
+    val rerouteState: RerouteState
+
+    /**
+     * Invoked whenever re-route is needed. As instance when a driver is off route. Called just after
+     * an off route event.
+     *
+     * @see [OffRouteObserver]
+     */
+    fun reroute(routesCallback: RoutesCallback)
+
+    /**
+     * Invoked when re-route is not needed anymore (as instance when driver returned to previous route).
+     * Might be ignored depends on [RerouteState] (when route has been fetched no sense to interrupt re-route)
+     */
+    fun interruptReroute()
+
+    /**
+     * Route Callback is useful to set new route(s) on reroute event. Doing the same as
+     * [MapboxNavigation.setRoutes].
+     */
+    interface RoutesCallback {
```

Comment:
This callback is now ignored. Since the `DirectionsSession` should be the source of truth, let's document it this way and remove this callback altogether.

Reply:
When user overwrite `RerouteController` and set custom realisation he should do use `RoutesCallback` to set new route(s). Default implementation ignore callback only because `DirectionsSession` update routes by it own

Reply:
In this case, if they are circumventing the directions session when making a request, they can then pass the direction session instance to their implementation and set the route whenever it's ready. We don't have to provide that abstraction.

Reply:
imo developer must know what sdk expects of him, so there's intuitively `after reroute happened you(developer) should do call RerouteCallback`

Reply:
They have to implement the callback themselves anyway, it's not handled by the SDK, so I think it defeats the purpose.

Reply:
what do you mean `it's not handled by the SDK`
https://github.com/mapbox/mapbox-navigation-android/pull/3184/files#diff-e3ca7b557efaeed833ca6f8f76dc2d03R605-R610

Reply:
Ah, that's a good point, I missed it. Although, now we're running into a risk where we are setting the route internally but if the developer actually used the `getRoutes` to fetch the route, we will duplicate the setting.

Reply:
great note! I think it's handled by interrupt reroute if `getRoute`/`setRoute` has been invoked